### PR TITLE
Implement faster FilteredRecursiveDirectoryIterator

### LINKED ISSUE #135: parallel-lint iterates directory structure even if excluded
I was wondering that parallel-lint works thru a deep directory structure (e.g.  `vendor/`) even though I passed `--exclude vendor` as a cli arg.

it seems the tools is deep iterating everything and does the filtering afterwards.

wouldn't it be way faster to ignore the excluded directories in the first place, so we don't need to scan the whole directory structure to finally decided there is nothing to lint inside?

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/php-parallel-lint/PHP-Parallel-Lint/issues/135

running parallel lint in https://github.com/FriendsOfREDAXO/rexstan with `vendor/bin/parallel-lint . --exclude vendor` takes `0.3 seconds` with this PR while it took `9 seconds` before the PR

### CODE REVIEW

**Comment by staabm** [comment]
@jrfnl anything I can do to move this forward?

**Comment by jrfnl** [comment]
@staabm Not really. I'll try to find some time to review this over the next few weeks. That should hopefully help, but this is, in my experience, generally speaking a slow moving repo, so patience is the keyword...

File: src/Iterators/FilteredRecursiveDirectoryIterator.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            $flags = \FilesystemIterator::KEY_AS_PATHNAME | \FilesystemIterator::CURRENT_AS_FILEINFO;
+        }
+
+        parent::__construct($path, $flags);
+    }
+
+    #[ReturnTypeWillChange]
+    public function hasChildren($allowLinks = false)
+    {
+        var_dump($this->getRealPath());
```
Comment (jrfnl) [inline src/Iterators/FilteredRecursiveDirectoryIterator.php:25]: Looks like there is a stray `var_dump()` here ?

**Comment by staabm** [comment]
Tbh I have implemented against 1.x - we need to verify it against 2.x (and see whether the speed improvements are still visible on 2.x)

**Comment by staabm** [comment]
let my clarify my last comment: the PR works in parallel-lint 1.x which I am currently using and it gives a decent speed boost, because without it, `vendor` files get linted even though I am passing `--exclude vendor`.

I did not verify the fix on parallel-int 2.x .. it looks like on 2.x the vendor is properly ignored even before this patch

File: src/Iterators/FilteredRecursiveDirectoryIterator.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            $flags = \FilesystemIterator::KEY_AS_PATHNAME | \FilesystemIterator::CURRENT_AS_FILEINFO;
+        }
+
+        parent::__construct($path, $flags);
+    }
+
+    #[ReturnTypeWillChange]
+    public function hasChildren($allowLinks = false)
+    {
+        if (in_array($this->getBasename(), $this->excluded)) {h
```
Comment (jrfnl) [inline src/Iterators/FilteredRecursiveDirectoryIterator.php:25]:
```suggestion
        if (in_array($this->getBasename(), $this->excluded)) {
```

Looks like a stray parse error made it into the patch ?

**Comment by staabm** [comment]
@jrfnl we just talked about this :-)

(It might no longer be relevant)

**Comment by staabm** [comment]
I can no longer reproduce the slownes in the latest 1.4.x release

**Pull request closed without merging**